base/loading.jl

```diff
@@ -408,9 +408,28 @@ end
 ## generic project & manifest API ##
 
 const project_names = ("JuliaProject.toml", "Project.toml")
-const manifest_names = ("JuliaManifest.toml", "Manifest.toml")
+const default_manifest_names = ("JuliaManifest.toml", "Manifest.toml")
 const preferences_names = ("JuliaLocalPreferences.toml", "LocalPreferences.toml")
 
+function manifest_names()
+    env_var_name = "JULIA_MANIFEST_NAME"
+    custom_def = strip(get(ENV, env_var_name, ""))
+    if !isempty(custom_def)
+        if !endswith(custom_def, ".toml")
+            throw(ArgumentError("Manifest name provided by `$env_var_name` must end with `.toml`. Provided: $(repr(custom_def))"))
+        end
+        man_name = expand_version_placeholders(custom_def)
+        if occursin("#", man_name)
+            throw(ArgumentError("""
+                Auto version population of `JULIA_MANIFEST_NAME` failed. Some `#` characters were not replaced: $(repr(name))
+                A maximum of three # chars may be used."""))
+        end
+        return (man_name, man_name)
```

Comment:
```suggestion
        return (man_name, default_manifest_names...)
```

Reply:
I disagree with this. If the user sets `JULIA_MANIFEST_NAME`, that should be the only manifest name that Julia uses.

Reply:
I guess we could have a syntax like this: `export JULIA_MANIFEST_NAME="Manifest.#.#.toml:"`, where the `:` at the end of the string means "please append the default manifest names to the end of the list".

But if the user does not provide the `:`, I think that we should not append the default manifest names.

Reply:
Yeah, that's exactly the kind of thorny question that I would like to avoid answering 😂 

Which is why I'd prefer that we just not have the append case. If the user sets `JULIA_MANIFEST_NAME`, presumably they know what manifest name they want to use.

Reply:
> If the user sets JULIA_MANIFEST_NAME, that should be the only manifest name that Julia uses.

That is going to be super annoying if you swap projects to a non-versioned one.

> In the append case, which name should Pkg use to create a new manifest?

`Manifest.toml`? I thought that wouldn't change with the current PRs.

Reply:
As long as there's a way to override that behavior (e.g. don't put `:` at the end of the environment variable),   I guess that's fine.

Reply:
Should the default be to append, and then you have to put a special character not to append? Or should the default be to not spend, and you have to put a special character to append?

Reply:
> > If the user sets JULIA_MANIFEST_NAME, that should be the only manifest name that Julia uses.
> 
> That is going to be super annoying if you swap projects to a non-versioned one.

I mean, this feature is completely opt-in, right? I think we can expect users that use this feature to be able to set and unset the environment variable appropriately.

Reply:
I did notice it meant Revise wasn't detected, because the global env used the normal manifest name. That's a little annoying

Reply:
Doesn't make a whole lot of sense to use this feature for global environments. The intended use case is applications.

I would recommend (in the documentation for this environment variable) that we tell users "it is recommended that you only use the `JULIA_MANIFEST_NAME`  environment variable if you are also using the `--project` command-line option and/or the `JULIA_PROJECT` environment variable.

Reply:
I wasn't. But my startup.jl tried to load Revise from the global env and failed to find the manifest so couldn't load revise

Reply:
Ohhhhhh.... the startup file! Hmmm.... that deserves some thought.

Reply:
I guess using `:` to append would solve the startup file Revise issue.

That's how we already do e.g. `DEPOT_PATH` and `LOAD_PATH`. If it ends in `:` we append the defaults. Otherwise we do not append the defaults.